### ISSUE TITLE
Race Condition in RNN Layers using (Recurrent-)Dropout

### ISSUE THREAD

**Issue by mergian** [keras-team/tf-keras#58]
**System information**.
- Have I written custom code (as opposed to using a stock example script provided in Keras): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): CentOS
- TensorFlow installed from (source or binary): PYPI
- TensorFlow version (use command below): v2.11.0
- Python version: 3.7

**Describe the problem**.

RNN layers using Dropout can be victim of a race condition when executed in graph mode. The problem comes from here:
https://github.com/keras-team/keras/blob/master/keras/backend.py#L5016-L5021

This is the outer loop for the RNNCells. It allows parallel execution. However, the dropout/recurrent_dropout masks get generated inside this loop. Depending on the parallel execution, one rand gets executed before the other, causing non-deterministic random numbers.

The tricky part is, that as in total the number of random calls is correct, the global state correctly gets incremented, so that consecutive executions of the RNN can be correct.

I therefore build a reproducer that has the rough structure:

```python
for o in range(1000):
	reset_seed()
	result = []
	for i in range(10):
		result.append(model(data, training=True))
	if reference == result:
		print('passed')
	else:
		print_differences()
```

See full code below. When I run it on my server I get 50-80 failures within 1000 iterations of the outer loop. There is no pattern when this error occurs, which matches my theory of the a race condition.

www.tensorflow.org/api_docs/python/tf/while_loop further explains that `For correct programs, while_loop should return the same result for any parallel_iterations > 0.` (although I think they mean > 1). This assumption is certainly violated for RNN that use dropout/recurrent_dropout as the state cannot be independently computed.

**Describe the current behavior**.
Results of running RNN with enabled dropout can generate wrong results.

**Describe the expected behavior**.
Results of RNNs should be identical when between different runs when properly seeded.

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**.

- Do you want to contribute a PR? (yes/no): yes
- If yes, please read [this page](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md) for instructions
- Briefly describe your candidate solution(if contributing):

Change `32` to `1` in https://github.com/keras-team/keras/blob/a63bef2ac504d651568c8800a9c1fcdcbd1cb41c/keras/backend.py#L5019 ;)

**Standalone code to reproduce the issue**.

```python
import tensorflow as tf
import numpy as np
from termcolor import colored

tf.keras.utils.set_random_seed(123)

# OPTIONS ----------------------------------------------------------------------
kwargs = {
	'units':				3,
	'dropout':				0.3,
	'recurrent_dropout':	0.7
}
runs			= 1000
batch_size		= 1
seq_len			= 2
use_graph_mode	= True
# /OPTIONS ---------------------------------------------------------------------

#-------------------------------------------------------------------------------
inp		= tf.keras.Input((seq_len, 3,))
rnn		= tf.keras.layers.SimpleRNN	(**kwargs)(inp)
gru		= tf.keras.layers.GRU		(**kwargs)(inp)
lstm	= tf.keras.layers.LSTM		(**kwargs)(inp)
model	= tf.keras.Model(inputs=[inp], outputs=[
	rnn,
	gru,
	lstm,
])

args = np.ones((batch_size, seq_len, 3,), np.float32)
reference = None
failed = 0

if use_graph_mode:  func = tf.function(model.__call__).get_concrete_function(args, training=True)
else:               func = model.__call__

for o in range(runs):
	print(f"Run #{o}: ", end='')

	tf.keras.utils.set_random_seed(123)

	results = {}
	for _ in range(10):
		res = func(args, training=True)
		output = model.output
		if not isinstance(res, (list, tuple)):
			output, res = [model.output], [res]

		for k, v in zip(output, res):
			results.setdefault(k.name, []).append(np.reshape(v.numpy(), -1))

	if reference is None:
		print('using as reference')
		reference = results
	else:
		def equal_list(A, B):
			assert len(A) == len(B)
			for a, b in zip(A, B):
				if not (a == b).all():
					return False
			return True

		def equal_dict(A, B):
			assert len(A) == len(B)
			for a, b in zip(A.values(), B.values()):
				if not equal_list(a, b):
					return False				
			return True

		if equal_dict(reference, results):
			print(colored('passed', 'green'))
		else:
			failed += 1
			print(colored('failed', 'red'))

			for (k, a), b in zip(reference.items(), results.values()):
				if not equal_list(a, b):
					print('\t', k)
					for i, (c, d) in enumerate(zip(a, b)):
						passed = (c == d).all()
						print(colored(f'\t\t{"" if passed else ">> "}#{i}: {c} {d}', 'green' if passed else 'red'))

print(f"## {failed}/{runs} failed!")
```

**Source code / logs**.

The `>>` indicate the line where the reference (left) does not match the current output (right).

```
Run #0: using as reference
Run keras-team/keras#1: passed
Run keras-team/keras#2: passed
Run keras-team/keras#3: passed
Run keras-team/keras#4: passed
Run keras-team/keras#5: passed
Run keras-team/keras#6: failed
         lstm/strided_slice_3:0
                #0: [-0.04476803  0.03647946  0.43398207] [-0.04476803  0.03647946  0.43398207]
                keras-team/keras#1: [-0.06646455  0.08540117  0.4261022 ] [-0.06646455  0.08540117  0.4261022 ]
                keras-team/keras#2: [-0.05694697 -0.03903455  0.31383437] [-0.05694697 -0.03903455  0.31383437]
                keras-team/keras#3: [-0.05401477  0.00309605  0.22946905] [-0.05401477  0.00309605  0.22946905]
                keras-team/keras#4: [-0.06022511  0.02117318  0.33324468] [-0.06022511  0.02117318  0.33324468]
                keras-team/keras#5: [ 0.0327878  -0.02470545  0.31550086] [ 0.0327878  -0.02470545  0.31550086]
                keras-team/keras#6: [-0.15186085  0.03566789  0.21747145] [-0.15186085  0.03566789  0.21747145]
                >> keras-team/keras#7: [-0.06862105  0.03145761  0.39546916] [-0.10350011  0.01307995  0.40501142]
                keras-team/keras#8: [-0.05030819  0.02323429  0.42084986] [-0.05030819  0.02323429  0.42084986]
                keras-team/keras#9: [-0.06160946  0.01366921  0.31102264] [-0.06160946  0.01366921  0.31102264]
Run keras-team/keras#7: passed
Run keras-team/keras#8: passed
Run keras-team/keras#9: passed
Run keras-team/keras#10: passed
Run keras-team/keras#11: passed
Run keras-team/keras#12: passed
Run keras-team/keras#13: passed
Run keras-team/keras#14: passed
Run keras-team/keras#15: passed
Run keras-team/keras#16: passed
Run keras-team/keras#17: passed
Run keras-team/keras#18: passed
Run keras-team/keras#19: passed
Run keras-team/keras#20: passed
Run keras-team/keras#21: passed
Run keras-team/keras#22: passed
Run keras-team/keras#23: passed
Run keras-team/keras#24: passed
Run keras-team/keras#25: passed
Run keras-team/keras#26: passed
Run keras-team/keras#27: passed
Run keras-team/keras#28: failed
         simple_rnn/strided_slice_3:0
                >> #0: [ 0.8512929  -0.7614667   0.11939589] [0.90060455 0.55757344 0.99093455]
                keras-team/keras#1: [-0.9390723  0.5154745  0.3419465] [-0.9390723  0.5154745  0.3419465]
                keras-team/keras#2: [-0.4668007  0.9807295 -0.526169 ] [-0.4668007  0.9807295 -0.526169 ]
                keras-team/keras#3: [ 0.9664895  -0.77306604  0.5689427 ] [ 0.9664895  -0.77306604  0.5689427 ]
                keras-team/keras#4: [ 0.84656304 -0.01103731  0.62985945] [ 0.84656304 -0.01103731  0.62985945]
                keras-team/keras#5: [ 0.84656304 -0.01103731  0.62985945] [ 0.84656304 -0.01103731  0.62985945]
                keras-team/keras#6: [ 0.9915579   0.55109423 -0.42523235] [ 0.9915579   0.55109423 -0.42523235]
                keras-team/keras#7: [-0.82279384  0.68517345  0.38375577] [-0.82279384  0.68517345  0.38375577]
                keras-team/keras#8: [0.94490975 0.5255155  0.9951209 ] [0.94490975 0.5255155  0.9951209 ]
                keras-team/keras#9: [0.98280257 0.66127515 0.99306947] [0.98280257 0.66127515 0.99306947]
Run keras-team/keras#29: passed
Run keras-team/keras#30: passed
Run keras-team/keras#31: passed
Run keras-team/keras#32: passed
...
```
